crates/bevy_pbr/src/light.rs

```diff
@@ -568,19 +784,135 @@ pub(crate) fn assign_lights_to_clusters(
+        };
+        let far_z = far_z.max(first_slice_depth);
+
         let cluster_factors = calculate_cluster_factors(
-            // NOTE: Using the special cluster near value
-            clusters.near,
-            camera.far,
-            clusters.axis_slices.z as f32,
+            first_slice_depth,
+            far_z,
+            cluster_dimensions.z as f32,
             is_orthographic,
         );
 
+        let max_indices = ViewClusterBindings::MAX_INDICES;
+
+        if max_indices
+            < (cluster_dimensions.x * cluster_dimensions.y * cluster_dimensions.z) as usize
+                * light_count
+        {
+            let mut cluster_index_estimate = 0.0;
+            for light in lights.iter() {
+                let light_sphere = Sphere {
+                    center: light.translation,
+                    radius: light.range,
+                };
+
+                // Check if the light is within the view frustum
+                if !frustum.intersects_sphere(&light_sphere) {
+                    continue;
+                }
+
+                // calculate a conservative aabb estimate of number of clusters affected by this light
+                // this overestimates index counts by at most 50% (and typically much less) when the whole light range is in view
+                // it can overestimate more significantly when light ranges are only partially in view
+                let (light_aabb_min, light_aabb_max) = cluster_space_light_aabb(
+                    is_orthographic,
+                    inverse_view_transform,
+                    camera.projection_matrix,
+                    &light_sphere,
+                );
+
+                // since we won't adjust z slices we can calculate exact number of slices required in z dimension
+                let z_cluster_min = view_z_to_z_slice(
+                    cluster_factors,
+                    cluster_dimensions.z as f32,
+                    light_aabb_min.z,
+                    is_orthographic,
+                );
+                let z_cluster_max = view_z_to_z_slice(
+                    cluster_factors,
+                    cluster_dimensions.z as f32,
+                    light_aabb_max.z,
+                    is_orthographic,
+                );
+                let z_count =
+                    z_cluster_min.max(z_cluster_max) - z_cluster_min.min(z_cluster_max) + 1;
+
+                // calculate x/y count using floats to avoid overestimating counts due to large initial tile sizes
+                let xy_min = light_aabb_min.xy();
+                let xy_max = light_aabb_max.xy();
+                // multiply by 0.5 to move from [-1,1] to [-0.5, 0.5], max extent of 1 in each dimension
+                let xy_count = (xy_max - xy_min)
+                    * 0.5
+                    * Vec2::new(cluster_dimensions.x as f32, cluster_dimensions.y as f32);
+
+                // add up to 2 to each axis to account for overlap
```

Comment:
```suggestion
                // add up to 2 to each axis to account for aabb overlap as the view space cluster aabbs
                // containing cluster frustum overlap each other
```

Reply:
this isn't actually aabb overlap, this is about the fact that a light with a range of epsilon can theoretically still touch 2x2 clusters if it happens to sit right on the border.